lib/monitor.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2002-2018 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "monitor.h"
+
+#if SYSLOG_NG_HAVE_INOTIFY
+static void
+handler(void *_fm, struct inotify_event *event)
+{
+  FileMonitor *fm = (FileMonitor *) _fm;
+
+  if (event->mask & IN_MODIFY)
+    fm->main_loop_reload_config(fm->self);
+
+  // 1- handle case where editing the file using vim causes an event 'IN_IGNORED'
+  //     and subsequently the watch gets deleted by ivykis: ivykis/src/iv_inotify.c:88
+  //   2- handle case when file is edited and contains a parse error, syslog-ng changes
+  //     file attributes and file becomes unreadable
+
+  //    - note : for case when file was deleted, syslog-ng would still attempt to reload,
```

Comment:
Maybe this case should be handled ? (deletion)
But I think the other use case is when a file is moved, as I think that is not handled correctly now.

Reply:
If the file is moved, the watch on the original file is eventually removed by the kernel so it's the same as if the file was deleted (nothing to watch for modifications - since the watch is not dynamic and will not track original file path changes). 

We can watch on the event ``IN_MOVED`` but it's not very useful since if user moved the original file back to its path, the watch would've been gone anyway and no way to re-watch other than restarting syslog-ng.